src/machine/machine_fe310.go

```diff
@@ -181,3 +182,147 @@ func (spi SPI) Transfer(w byte) (byte, error) {
 	// return data
 	return byte(spi.Bus.RXDATA.Get() & sifive.QSPI_RXDATA_DATA_Msk), nil
 }
+
+// I2C on the FE310-G002.
+type I2C struct {
+	Bus *sifive.I2C_Type
+}
+
+// I2CConfig is used to store config info for I2C.
+type I2CConfig struct {
+	Frequency uint32
+	SCL       Pin
+	SDA       Pin
+}
+
+// Configure is intended to setup the I2C interface.
+func (i2c I2C) Configure(config I2CConfig) error {
+
+	var clockFrequency uint32 = 32000000
+
+	if config.Frequency == 0 {
+		config.Frequency = TWI_FREQ_100KHZ
+	}
+
+	if config.SDA == 0 && config.SCL == 0 {
+		config.SDA = I2C0_SDA_PIN
+		config.SCL = I2C0_SCL_PIN
+	}
+
+	var prescaler = clockFrequency/(5*config.Frequency) - 1
+
+	// disable controller before setting the prescale registers
+	i2c.Bus.CTR.ClearBits(sifive.I2C_CTR_EN)
+
+	// set prescaler registers
+	i2c.Bus.PRER_LO.Set(uint32(prescaler & 0xff))
+	i2c.Bus.PRER_HI.Set(uint32((prescaler >> 8) & 0xff))
+
+	// enable controller
+	i2c.Bus.CTR.SetBits(sifive.I2C_CTR_EN)
+
+	config.SDA.Configure(PinConfig{Mode: PinI2C})
+	config.SCL.Configure(PinConfig{Mode: PinI2C})
+
+	return nil
+}
+
+// Tx does a single I2C transaction at the specified address.
+// It clocks out the given address, writes the bytes in w, reads back len(r)
+// bytes and stores them in r, and generates a stop condition on the bus.
+func (i2c I2C) Tx(addr uint16, w, r []byte) error {
+	var err error
+	if len(w) != 0 {
+		// send start/address for write
+		i2c.sendAddress(addr, true)
+
+		// ACK received (0: ACK, 1: NACK)
+		if i2c.Bus.CR_SR.HasBits(sifive.I2C_SR_RX_ACK) {
+			return errors.New("I2C write error: expected ACK not NACK")
+		}
+
+		// write data
+		for _, b := range w {
+			err = i2c.writeByte(b)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if len(r) != 0 {
+		// send start/address for read
+		i2c.sendAddress(addr, false)
```

Comment:
This might be problematic for some I2C devices that expect the write and read to happen in a single transaction.
This doesn't need fixing right away but something to be aware of if some hardware fails.

Reply:
I am not sure to understand why that wouldn't be the case, the stop signal to end the transaction is only generated [after](https://github.com/tinygo-org/tinygo/pull/982/files#diff-8a77a1909742be6b60f3ea35660048ddR276) the write and read.

When calling `i2c.sendAddress()` the second time a repeated start is generated but this is still the same transaction.

Reply:
Ah I see. Well I hope devices can deal with repeated starts, then (you'd be surprised how many are slightly non conforming).